File: Python/RSAEncryption/RSAEncryption.py
"""
    Master Class

    make new instance of this class to make RSA Encryption work
"""

import sys
import os

path = os.getcwd() + '/RSAEncryption/public_key_private_key/'
pathMath = os.getcwd() + '/RSAEncryption/math_operation/'
path_encrypt_decrypt = os.getcwd() + '/RSAEncryption/encrypt_decrypt/'
sys.path.append(os.path.abspath(path))
sys.path.append(os.path.abspath(pathMath))
sys.path.append(os.path.abspath(path_encrypt_decrypt))

from RSAEncryption.public_key_private_key.public_private_key_generator import *
from RSAEncryption.math_operation.primeNumberOp import *
from RSAEncryption.encrypt_decrypt.DecryptRSA import *
from RSAEncryption.encrypt_decrypt.EncryptRSA import *


class RSA:

    """
        Set the range of prime number from min to max
    """
    def __init__(self, minPrimeNumber, maxPrimeNumber):
        self.__pValue = generatePrimeNumber(minPrimeNumber, maxPrimeNumber)
        self.__qValue = generatePrimeNumber(minPrimeNumber, maxPrimeNumber)

        while self.__qValue==self.__pValue:
            self.__qValue = generatePrimeNumber(minPrimeNumber, maxPrimeNumber)

        self.n_value = self.__calculate_n()
        self.__x_of_n = self.__calculate_x_of_n()
        self.public_key = self.__calculate_public_key()
        self.__private_key = self.__calculate_private_key()
    
    # Public methods =================================================================
    def get_n_value(self):
        return self.n_value

    def get_x_of_n(self):
        return self.__x_of_n

    def get_public_key(self):
        return self.public_key

    def __get_private_key(self):
        return self.__private_key

    """
        Return array of encrypted chunks of data
        use '|'.join(iterable) to make it converted to string with '|' for data separator

        EXAMPLE USAGE:

        encrypted = rsa_object_instance.encrypt(text)

        print(encrypted) <--- this will print in the form of array

        This code below will print in the form of string with '|' separator
        print('|'.join(str(data) for data in encrypted))

        #### OR you can store it in variable ####

        EXAMPLE:

        encrypted = '|'.join(str(data) for data in encrypted)
    """
    def encrypt(self, text):
        return encrypt_data(self.public_key, self.n_value, text)

    """
        Return array of decrypted characters

        same usage as encrypt method
    """
    def decrypt(self, data):
        return decrypt_data(data, self.__private_key, self.n_value)

    # Private methods ================================================================
    def __calculate_n(self):
        return (self.__pValue * self.__qValue)
    
    def __calculate_x_of_n(self):
        return find_x_of_n(self.__pValue, self.__qValue)

    def __calculate_public_key(self):
        return generate_public_key(self.get_x_of_n())

    def __calculate_private_key(self):
        return generate_private_key(self.get_x_of_n(), self.get_public_key())

File: Javascript/RSAEncryption/RSAEncryption.js
/*
    This is the main class for RSA Encryption Javascript module

    © Author Fredo Ronan
*/

import { generate_prime } from './prime-number-op.js';
import { find_x_of_n, generate_private_key, generate_public_key } from './public_key_private_key.js';
import { decrypt_data, encrypt_data } from './encrypt-decrypt.js';


export class RSAEncryption {

    #p_value;
    #q_value;
    #n_value;
    #x_of_n;
    #private_key;
    #public_key;

    constructor(minPrimeNumber, maxPrimeNumber){
        this.#p_value = generate_prime(minPrimeNumber, maxPrimeNumber);
        this.#q_value = generate_prime(minPrimeNumber, maxPrimeNumber);

        while(this.#q_value===this.#p_value){
            this.#q_value = generate_prime(minPrimeNumber, maxPrimeNumber);
        }

        this.#n_value = this.#calculate_n();
        this.#x_of_n = this.#calculate_x_of_n();
        this.#public_key = this.#calculate_public_key();
        this.#private_key = this.#calculate_private_key();
    }

    // Public Methods ================================================================
    get_public_key(){
        return this.#public_key;
    }

    get_n_value(){
        return this.#n_value;
    }


    // ENCRYPTION SIDE //////////////////////////////////////////////////////////////
    encrypt_array_result(text){
        return encrypt_data(this.#public_key, this.#n_value, text);
    }

    /* 
        To encrypt with the return of string with separator

        Example:

        encrypt_string_result(message, '|');
        That code will return the result like this ===> 8298|19209|910241|83982|8283|92763| (just example) 
    */
    encrypt_string_result(text, separator){
        let temp_encrypted = encrypt_data(this.#public_key, this.#n_value, text);
        let encrypted_data = temp_encrypted.join(separator);
        encrypted_data += separator;
        return encrypted_data;
    }


    // DECRYPTION SIDE //////////////////////////////////////////////////////////////
    decrypt_array_result(encrypted_data){
        return decrypt_data(encrypted_data, this.#private_key, this.#n_value, 'NULL');
    }

    decrypt_string_result(encrypted_data, separator){
        let temp_decrypted = decrypt_data(encrypted_data, this.#private_key, this.#n_value, separator);
        let decrypted_data = temp_decrypted.join('');
        return decrypted_data;
    }


    // Private Methods ================================================================
    #calculate_n(){
        return (this.#p_value * this.#q_value);
    }

    #calculate_x_of_n(){
        return find_x_of_n(this.#p_value, this.#q_value);
    }

    #calculate_public_key(){
        return generate_public_key(this.#x_of_n);
    }

    #calculate_private_key(){
        return generate_private_key(this.#x_of_n, this.#public_key);
    }
}